zerver/views/streams.py

```diff
@@ -384,19 +386,27 @@ def update_stream_backend(
     if stream_post_policy is not None:
         do_change_stream_post_policy(stream, stream_post_policy, acting_user=user_profile)
 
-    if can_remove_subscribers_group_id is not None:
-        if sub is None and stream.invite_only:
-            # Admins cannot change this setting for unsubscribed private streams.
-            raise JsonableError(_("Invalid stream ID"))
-
-        user_group = access_user_group_for_setting(
-            can_remove_subscribers_group_id,
-            user_profile,
-            setting_name="can_remove_subscribers_group",
-            require_system_group=True,
-        )
-
-        do_change_can_remove_subscribers_group(stream, user_group, acting_user=user_profile)
+    request_settings = {
+        k: v for k, v in list(locals().items()) if (k in Stream.stream_permission_group_settings)
+    }
+    for setting_name, setting_value in list(request_settings.items()):
+        if setting_value is not None and getattr(stream, setting_name) != setting_value:
+            if sub is None and stream.invite_only:
+                # Admins cannot change this setting for unsubscribed private streams.
+                raise JsonableError(_("Invalid stream ID"))
+
+            group_permission_setting_info = Stream.stream_permission_group_settings[setting_name]
```

Comment:
Would it make more sense to just loop over `Stream.stream_permission_group_settings.items()` once, and just have a `if not hasattr(locals(), setting_group_id_name): continue` check at the top of the loop? 

I feel like having to reference this again in order to get the metadata feels a bit harder to read than `for setting_name, permissions_configuration in Stream.stream_permission_group_settings:` as the outer loop here.

Reply:
The updated loop looks good; added this hunk to avoid the extra fetch:

```
diff --git a/zerver/views/streams.py b/zerver/views/streams.py
index d94914dbcb..7b98b0ec2c 100644
--- a/zerver/views/streams.py
+++ b/zerver/views/streams.py
@@ -399,14 +399,13 @@ def update_stream_backend(
                 raise JsonableError(_("Invalid stream ID"))
 
             user_group_id = request_settings_dict[setting_group_id_name]
-            group_permission_setting_info = Stream.stream_permission_group_settings[setting_name]
             user_group = access_user_group_for_setting(
                 user_group_id,
                 user_profile,
                 setting_name=setting_name,
-                require_system_group=group_permission_setting_info.require_system_group,
-                allow_internet_group=group_permission_setting_info.allow_internet_group,
-                allow_owners_group=group_permission_setting_info.allow_owners_group,
+                require_system_group=permissions_configuration.require_system_group,
+                allow_internet_group=permissions_configuration.allow_internet_group,
+                allow_owners_group=permissions_configuration.allow_owners_group,
             )
             do_change_stream_group_based_setting(
                 stream, setting_name, user_group, acting_user=user_profile
```